NuRadioReco/modules/io/LOFAR/readLOFARData.py

```diff
@@ -0,0 +1,479 @@
+                # Metadata is a list containing:
+                # station name, antenna set, tbb timestamp (seconds), tbb timestamp (nanoseconds),
+                # station clock frequency (Hz), positions of antennas, dipole IDs and calibration delays per dipole
+
+    @register_run()
+    def run(self, detector, trace_length=65536):
+        """
+        Runs the reader with the provided detector. For every station that has files associated with it, a Station
+        object is created together with its channels (pulled from the detector description). Every channel also gets
+        a group ID which corresponds to the polarisation (i.e. 0 for even and 1 for odd), as to be able to retrieve
+        all channels per polarisation during processing.
+
+        Parameters
+        ----------
+        detector: Detector object
+            The detector description to be used for this event.
+        trace_length: int
+            Desired length of the trace to be loaded from TBB files.
+
+        Yields
+        ------
+        evt: Event object
+            The event containing all the loaded traces.
+        """
+        # Create an empty with 1 run, as only 1 shower per event
+        evt = NuRadioReco.framework.event.Event(1, self.__event_id)
+
+        # Add all Detector stations to Event
+        for station_name, station_dict in self.__stations.items():
+            station_id = int(station_name[2:])
+            station_files = station_dict['files']
+
+            if len(station_files) == 0:
+                continue
+            station = NuRadioReco.framework.station.Station(station_id)
+
+            # Use KRATOS io functions to access trace
+            lofar_trace_access = getLOFARtraces(
+                station_files,
+                self.meta_dir,
+                self.__lora_timestamp,
+                self.__lora_timestamp_ns,
+                trace_length
+            )
+            channels_deviating, channels_missing_counterpart = lofar_trace_access.check_trace_quality()
+
+            self.logger.debug("Channels deviating: %s" % channels_deviating)
+            self.logger.debug("Channels no counterpart: %s" % channels_missing_counterpart)
+
+            # done here as it needs median timing values over all traces in the station
+            flagged_channel_ids = channels_deviating.union(channels_missing_counterpart)
+            for channel_id in detector.get_channel_ids(station_id):
+                if channel_id in flagged_channel_ids:
+                    continue                 
+                if detector.get_channel(station_id, channel_id)['ant_orientation_phi'] == 225.0:
+                    channel_group = 0
+                elif detector.get_channel(station_id, channel_id)['ant_orientation_phi'] == 135.0:
+                    channel_group = 1
+                else:
+                    raise ValueError('Orientation not implemented')
```

Comment:
I think this came up in the call - this looks a bit risky to me, probably safer/more natural to implement this in the detector descriptions. E.g. what if a station is slightly differently oriented, or you end up calibrating the antenna orientations so they're no longer exactly 225 degrees, etc.

Reply:
Indeed, we should add this to the Detector description (also changing the interpretation of the channel_group_id as discussed below). I'll make this a separate issue to implement in a different branch, I think it would be too messy to add this through this branch.